### ISSUE TITLE
CI: Conformance Ginkgo - f01-agent-chaos

### ISSUE THREAD

**Issue by giorio94** [cilium/cilium#29236]
## CI failure

Hit on https://github.com/cilium/cilium/pull/29232
Link: https://github.com/cilium/cilium/actions/runs/6901702835/job/18777015895
Sysdump: [K8sAgentChaosTest_Connectivity_demo_application_Endpoint_can_still_connect_while_Cilium_is_not_running.tar.gz](https://github.com/cilium/cilium/files/13390596/K8sAgentChaosTest_Connectivity_demo_application_Endpoint_can_still_connect_while_Cilium_is_not_running.tar.gz)

```
• Failure in Spec Setup (BeforeEach) [502.846 seconds]
K8sAgentChaosTest
/home/runner/work/cilium/cilium/test/ginkgo-ext/scopes.go:461
  Connectivity demo application
  /home/runner/work/cilium/cilium/test/ginkgo-ext/scopes.go:461
    Endpoint can still connect while Cilium is not running [BeforeEach]
    /home/runner/work/cilium/cilium/test/ginkgo-ext/scopes.go:515

    Pods are not ready after timeout
    Expected
        <*errors.errorString | 0xc002a4c7d0>: 
        timed out waiting for pods with filter -l zgroup=testDS to be ready: 4m0s timeout expired
        {
            s: "timed out waiting for pods with filter -l zgroup=testDS to be ready: 4m0s timeout expired",
        }
    to be nil
```

From a quick look at the sysdump it seems that one of the cilium agents is unhealthy, and the endpoints API is stuck. For instance, there are the following in the CNI logs:

```
level=warning msg="Errors encountered while deleting endpoint" error="Cilium API client timeout exceeded" subsys=cilium-cni
level=warning msg="Errors encountered while deleting endpoint" error="Cilium API client timeout exceeded" subsys=cilium-cni
level=warning msg="Unable to create endpoint" containerID=13d1b30b975e278fe37a27eeb18527bf3b5d191f5b0cebb952400bd773759575 error="Cilium API client timeout exceeded" eventUUID=06e3591e-8039-4b5b-99cd-05d438cf686d subsys=cilium-cni
```

There are suspicious gops traces suggesting a possible deadlock around endpoints manager, e.g.,:

```
goroutine 73 [sync.RWMutex.RLock, 4 minutes]:
sync.runtime_SemacquireRWMutexR(0xc002e6fbe8?, 0x69?, 0xc0000b6120?)
	/usr/local/go/src/runtime/sema.go:82 +0x25
sync.(*RWMutex).RLock(...)
	/usr/local/go/src/sync/rwmutex.go:71
github.com/cilium/cilium/pkg/endpointmanager.(*endpointManager).GetEndpoints(0xc000f898c0)
	/go/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:589 +0x45
github.com/cilium/cilium/pkg/endpointmanager.(*endpointManager).RegenerateAllEndpoints(0xc000adb790?, 0xc002eb8d50)

goroutine 241 [sync.RWMutex.RLock, 4 minutes]:
sync.runtime_SemacquireRWMutexR(0xc003242fc0?, 0x85?, 0xc00334ede0?)
	/usr/local/go/src/runtime/sema.go:82 +0x25
sync.(*RWMutex).RLock(...)
	/usr/local/go/src/sync/rwmutex.go:71
github.com/cilium/cilium/pkg/endpointmanager.(*endpointManager).GetPolicyEndpoints(0xc000f898c0)
	/go/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:601 +0x45
github.com/cilium/cilium/daemon/cmd.(*Daemon).policyAdd(0xc000c8a000, {0xc001275da0?, 0x1, 0x1}, 0xc002b15c80, 0xc002b15ce0?)
	/go/src/github.com/cilium/cilium/daemon/cmd/policy.go:281 +0x66c
github.com/cilium/cilium/daemon/cmd.(*PolicyAddEvent).Handle(0xc00235d7d0?, 0x6c65640a2c737473?)
```

**Comment by marseel** [comment]
Seems like this is also the reason for sig-network kind tests to be failing, got a similar deadlock in the stack there:
```
goroutine 79 [sync.RWMutex.RLock, 16 minutes]:
sync.runtime_SemacquireRWMutexR(0xc005787488?, 0x85?, 0xc00357dc80?)
        /usr/local/go/src/runtime/sema.go:82 +0x25
sync.(*RWMutex).RLock(...)
        /usr/local/go/src/sync/rwmutex.go:71
github.com/cilium/cilium/pkg/endpointmanager.(*endpointManager).GetPolicyEndpoints(0xc0011761b0)
        /go/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:601 +0x45
github.com/cilium/cilium/daemon/cmd.(*Daemon).policyDelete(0xc001e9ca80, {0xc00270f440?, 0x4, 0x4}, 0xc0014607e0, 0xc005787ea8?)
        /go/src/github.com/cilium/cilium/daemon/cmd/policy.go:529 +0x20d
github.com/cilium/cilium/daemon/cmd.(*PolicyDeleteEvent).Handle(0xc001ba9f50?, 0xc0016bb860?)
        /go/src/github.com/cilium/cilium/daemon/cmd/policy.go:481 +0x2f
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run.func1()
        /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:244 +0x133
sync.(*Once).doSlow(0x0?, 0x0?)
        /usr/local/go/src/sync/once.go:74 +0xbf
sync.(*Once).Do(...)
        /usr/local/go/src/sync/once.go:65
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run(0x0?)
        /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:232 +0x3c
created by github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).Run in goroutine 1
        /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:228 +0x69
```

**Comment by marseel** [comment]
ok, that was a tough one.

So what is happening here:
- goroutine X acquires RLock [here](https://github.com/cilium/cilium/blob/bacf2e8718e603b6b74bd599711d73f3f7bfcddb/pkg/endpointmanager/manager.go#L417)
- routine Y tries to get `Lock` (doesn't matter reason, for example can be `markAndSweep`)
- Starting now, no one can acquire any lock (neither Lock or RLock) as there is a write lock waiting so read lock will block from now on too.
- goroutine X tries to send event [here](https://github.com/cilium/cilium/blob/bacf2e8718e603b6b74bd599711d73f3f7bfcddb/pkg/endpointmanager/manager.go#L419)
- authmap_gc tries to acquire RLock [here](https://github.com/cilium/cilium/blob/bacf2e8718e603b6b74bd599711d73f3f7bfcddb/pkg/auth/authmap_gc.go#L427)
- Deadlock, as one goroutine is waiting on write lock, we cannot acquire read lock :skull: 

So basically [this comment](https://github.com/cilium/cilium/blob/8402fd31a40c5c7d3db34bd8ff00bdf5c94a6fb8/pkg/endpointmanager/subscribe.go#L9-L19) is incorrect, because we can't even do read operations on endpointManager :cry:

**Comment by marseel** [comment]
Ah, looks like it's already being addressed by https://github.com/cilium/cilium/issues/29078

**Comment by marseel** [comment]
I will close this issue in favor of the other one and also mark the other one as a release-blocker.
